Rephrase documentation about literals and rational types. (#12845)
ethereum/solidity: docs/types/value-types.rst

In Solidity, division rounds towards zero. This means that ``int256(-5) / int256(2) == int256(-2)``.

Note that in contrast, division on :ref:`literals<rational_literals>` results in fractional values

> of arbitrary precision. Here, the word "arbitrary" means "any" or "unlimited."

Reviewer: As pointed out by @ekpyron this is really idiomatic and explaining the word here feels just weird to me. How about we just link to Wikpiedia for those who are not familiar with the term? I think this would be less obtrusive for others, who are:

```suggestion
of `arbitrary precision <https://en.wikipedia.org/wiki/Arbitrary-precision_arithmetic>`_.
```
```
```
v

There is no additional semantic meaning added to a number literal containing underscores,
the underscores are ignored.

Number literal expressions retain unlimited precision until they are converted to a non-literal type (i.e. by
using them together with anything other than a number literal expression (like boolean literals) or by explicit conversion).
This means that computations do not overflow and divisions do not truncate

> in number literal expressions. This is achieved by representing fractional rational numbers as quotients of integers, and

Reviewer 2: Is it important to document how we do it internally? I'm asking because we need to find this sentence and modify it whenever we do a change.

Author: Fair enough, however that is the main point of this whole topic. I.e., that fractional rational numbers are represented as fractional rational numbers (which specifically means as quotients of integers. In other words, if that implementation changed, this whole section and every mention of "unlimited precision" (or "arbitrary precision") in the docs would probably have to be revisited.

Reviewer: I also think this is unnecessary. It's a compiler implementation detail not necessary to understand the semantics of literals in Solidity.

    Division on integer literals used to truncate in Solidity prior to version 0.4.0, but it now converts into a rational number, i.e. ``5 / 2`` is not equal to ``2``, but to ``2.5``.

.. note::
    In solidity, rational numbers and integers expressed as literals are of number literal type.
    Moreover, all number literal expressions (i.e. the expressions that
    contain only number literals and operators) are of number literal

>     type.  The number literal expressions ``1 + 2`` and ``2 + 1`` are both

Reviewer 2: I think it's important her eto state that it is exactly the same type.

Author: Ok. As a new reader the words "belong to the same number literal type" in the context of the rest of the original wording did not actually convey "that it is exactly the same type". Instead, the picture that was formed in my head, reading the original doc, is that "there is a set of number literal types which are fungible/compatible with each other, and that 1+2 and 2+1 could be of one of those types, and 3 of another of those types, but all of them are in the same family.

So, if it is true that they are literally "the one and the same type" the wording can be even more improved from my suggestion. But, in my experiments, I found that some implementations (foundry) seems to treat a fractional rational literal as one type, and a non-fractional rational literal as another type, but that those types are "fully compatible" with each other.

That's why I chose this wording.

Reviewer: I think you misunderstood what this paragraph is talking about see: https://github.com/ethereum/solidity/pull/12845#discussion_r845359000

Which, on the other hand, does indicate that we're not clear enough and it should be reworded. Just not the way it's currently in the PR :)

>     of number literal type, just the same as ``3``. There are number literal subtypes, such

Reviewer 2: Why is it a "subtype" and what do you mean by the term?

    as those for fractional rational numbers like ``1.5``.

.. note::
    Number literal expressions are converted into a non-literal type as soon as they are used with non-literal
    expressions. Disregarding types, the value of the expression ``2.5 + a + 0.5`` where ``a=1`` would be the

>     integer ``4``. However, ``a`` is of type ``uint128``. Therefore, an implicit conversion to a non-literal

Reviewer 2: What I meant with "Disregarding types" was that the value that is assigned to `b` is always an integer, regardless of the type of `a` or of any other sub-expression in the assigned value. What we can maybe say is that the expression `2.5 + 1 + 0.5` evaluates to `4` (if you think that this is not clear to the reader).

    Division on integer literals used to truncate in Solidity prior to version 0.4.0, but it now converts into a rational number, i.e. ``5 / 2`` is not equal to ``2``, but to ``2.5``.

.. note::

>     In solidity, rational numbers and integers expressed as literals are of number literal type.

Reviewer: Your version as a different meaning. The original sentence says that e.g. `1` and `2` belong to different types (called `int_const 1` and `int_const 2`; you can't use these types for variables but you can sometimes see them in error messages), while `1` and `1.0` are of the same type (both are `int_const 1`).

The important takeaway is that there are no separate types for integers and fractional numbers used in literals. In the AST `1.0` is indistinguishable from `1`. That distinction only appears when you convert the literal expression to a concrete integer or fixed-point type (note also that fixed-point types are currently not fully implemented).

    Moreover, all number literal expressions (i.e. the expressions that
    contain only number literals and operators) are of number literal
    type.  The number literal expressions ``1 + 2`` and ``2 + 1`` are both
    of number literal type, just the same as ``3``. Technically speaking, the internal representation
    for fractional rational numbers like ``1.5`` is different than that of ``3``, specifically to support

>     arbitrary precision. However, all number literals are treated as effectively the same type.

Reviewer: It's not different. They're both internally represented using the same multi-precision type. If you took the same number as a fractional vs integer (e,g `3.0` vs `3`) there would be absolutely no difference internally. You can consider them not just convertible but exactly equivalent for all intents and purposes is Solidity.

This is really getting into compiler's implementation details that are not relevant to understanding semantics of literals. I'd remove this part.

.. note::
    Number literal expressions are converted into a non-literal type as soon as they are used with non-literal
    expressions. If not for type conversion considerations, when ``a=1`` the value of the expression ``2.5 + a + 0.5``

>     would be the integer ``4``. However, ``a`` is of type ``uint128``. Therefore, an implicit conversion to a non-literal

Reviewer: Technically, it could be a literal type if a common literal type existed.
```suggestion
    would be the integer ``4``. However, ``a`` is of type ``uint128``. Therefore, an implicit conversion to a common
```

>     type must occur when evaluating the first part of the expression ``2.5 + a``. Since there is no such conversion

Reviewer: ```suggestion
    type must occur when evaluating the first part of the expression (``2.5 + a``). Since there is no such conversion
```

>     between the types of ``2.5`` (a fractional rational literal number type) and ``uint128`` (an unsigned integer

Reviewer: There are no fractional rational number types distinct from integer rational number types. Integer and fractional literals belong to the same rational type if their values are equal.
```suggestion
    between the types of ``2.5`` (a rational literal number type) and ``uint128`` (an unsigned integer
```

>     value type), the Solidity compiler does not accept this code.  See :ref:Operators<operator>

Reviewer: ```suggestion
    value type), the Solidity compiler does not accept this code.  See :ref:`Operators<operator>`.
```

>     between the types of ``2.5`` (a fractional rational literal number type) and ``uint128`` (an unsigned integer
>     value type), the Solidity compiler does not accept this code.  See :ref:Operators<operator>

Reviewer: It sounds as if you're talking about the type of `uint128` (i.e. type of a type), while you should be talking about the type of `a`.
```suggestion
    between the type of ``2.5`` (a fractional rational literal number type) and ``uint128``
    the Solidity compiler does not accept this code.  See :ref:Operators<operator>
```
